Content filter: try parsing the HTML before a DB query

joemcgill opened the pull request:
This is another approach to a content filter based off the work in #170.

This filter only matches images with a `src` including the path to our uploads directory to slim down on the number of images it attempts to parse. Then, we first try to determine the post ID and size by parsing the value of the class attribute and finally, run a database query only as a last resort.

This still needs some perf testing, but I wanted to get feedback on the approach.

joemcgill commented:
I think there's something wrong with the filter test, and not the code, but will confirm.

joemcgill commented:
@jaspermdegroot I figured out what went wrong with the tests and went back and brought in a few more of your original ideas from #144. Mainly, it's possible for the order of the classes to be different depending on which image size you use (which is strange, but so it is).

jaspermdegroot commented:
@joemcgill 

I like the approach of first trying to get the ID and size name from the classes in the image markup and only use the image url as fallback. I have written the reasons why I am in favor of this in a comment on the "main" PR for content filtering (https://github.com/ResponsiveImagesCG/wp-tevko-responsive-images/pull/170#issuecomment-141646811) to keep the discussion at one place.

When I first looked at your code I had some concerns about the simplified way to get the ID and size name from the classes, but you already addressed those. I will review the code again and add inline comments.

I already applied your change in the unit test to the content-filtering branch (https://github.com/ResponsiveImagesCG/wp-tevko-responsive-images/commit/7c7c775678c49cc73a77cd13dc707ac508f18f53) because I was curious if the tests would pass, but Travis seems to have issues at the moment so I still don't know.
We can squash your commits and then rebase against the content-filtering branch to make the commit that we will be merging only contain changes in the code itself.

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -18,7 +18,6 @@
 // Don't load the plugin directly
 defined( 'ABSPATH' ) or die( "No script kiddies please!" );
 
-// List includes
```
The comment shouldn't be removed

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
-		'tevkori_filter_content_images_callback',
-		$content
-	);
+	// Only match images in our uploads directory
```
Coding standards: end comment with period

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+	$uploads_dir = wp_upload_dir();
+	$path_to_upload_dir = $uploads_dir['baseurl'];
+
+	$content = preg_replace_callback( '|<img ([^>]+' . $path_to_upload_dir . '[^>]+) />|', '_tevkori_filter_content_images_callback', $content );
```
Coding standards: use multiple lines

jaspermdegroot replied:
```|<img ([^>]+' . $path_to_upload_dir . '[^>]+) />|```

The slash before the closing tag is optional in html5 so it would be good if the regex reflects that in case there are editors/page builders that omit it.
I think the space between ```)``` and ```/``` should be removed.
Might be good to add the ```i```modifier to make it case insensitive.

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+function _tevkori_filter_content_images_callback( $image ) {
+	list( $image_html, $atts ) = $image;
+	$id = $size = false;
+
```
We should bail here if there is already a ```srcset``` attribute in the markup.

```if ( false !== strpos( 'srcset="', $atts ) ) { ...}```

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+
+	// Grab ID and size info from core classes.
+	if ( preg_match( '/wp-image-([0-9]+)/i', $atts, $class_id ) ) {
+		$id = $class_id[1];
```
cast $id as integer

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+	// Grab ID and size info from core classes.
+	if ( preg_match( '/wp-image-([0-9]+)/i', $atts, $class_id ) ) {
+		$id = $class_id[1];
+	}
```
add empty line

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+	}
+
+	// If attempts to get values for `id` and `size` failed, use the
+	// src to query for matching values in `_wp_attachment_metadata`
```
Coding standards for multi line comments https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#5-2-multi-line-comments.
Period at end of last lne.

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
-			if ( $srcset_string ) {
-				$srcset = ' ' . $srcset_string;
+		// Query the DB to get the post id and meta values for any attachment
+		// containing the file name of our url.
```
Coding standards for multi line comments https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#5-2-multi-line-comments

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+				$size = 'full';
+			} else {
+				// Otherwise, we loop through the sizes until we find the one whose
+				// file name matches the file name of our image.
```
Coding standards for multi line comments

jaspermdegroot commented on wp-tevko-responsive-images.php:
```diff
@@ -308,68 +307,86 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
+		// If the query is successful, we can determine the ID and size.
+		if ( is_object( $meta_object ) ) {
+			$id = $meta_object->post_id;
+			// Unserialize the meta_value returned in our query.
```
empty line before the comment

jaspermdegroot commented:
@joemcgill 

At first I didn't understand why you didn't need to use a regex to strip the size part from the url but now I have taken a closer look I get it. So your solution is even better than I thought.
I tested the fallback, also with an editted image, and it works great!

Nice work!!

joemcgill commented:
@jaspermdegroot I think I've addressed all the issues specific to the content filter. If you agree, I'll squash all of this into a single commit and resolve conflicts with the original content filtering branch so we can merge this into that branch. Once we've done so, we can do any final cleanup before merging into dev.

jaspermdegroot commented:
@joemcgill 

Sounds good!
The only thing I didn't see in the code is the early check if the image already has a ```srcset```. Or did I miss something? Just want to make sure we don't forget to implement that.

joemcgill commented:
Oh no, good point. I'll add the early rip cord when a srcset is already present.

joemcgill commented:
I just finished doing some performance profiling, using similar methods to what @jaspermdegroot used in #144. The test site is running locally in a VM running VVV.

Each test shows two results (times in milliseconds).
A: w/o content filter (average : median)
B: with content filter (average : median)

## Single Image 225 words
A: 64 : 63
B: 69 : 66

## 6 images 2000 words (mixed filter methods)
A: 73.438 : 69
B: 89.5 : 85

## 20K words, 20 images using standard WP markup
A: 133.355 : 121
B: 197.688 : 185

##  20K words, 20 images with WP classes stripped (forcing a postmeta query)
A: 149.556 : 116
B: 225.552 : 191

joemcgill commented:
Replaced by #177
